### ISSUE TITLE
PDF Rendering issues

### ISSUE THREAD

**Issue by DorianRudolph** [pdf-rs/pdf_render#14]
Hey, 

I just tried this out with a random PDF from my download folder, and there seem to be a couple of issues: https://leica-camera.com/sites/default/files/pm-73760-Leica-SL2_Technical-Data_en.pdf

First page (pdf_render, evince):

![image](https://user-images.githubusercontent.com/10325838/208740490-279d2ea5-ac2b-408d-96e3-5f3e3dbf5dc1.png)

![image](https://user-images.githubusercontent.com/10325838/208740519-14787efe-1756-427e-ae80-e8a982d6f81b.png)

The first issue is in the title. Analyzing the PDF in Master PDF editor, one can see that there is the text "DAS WESENTLICHE" but with clipping path and white font behind the "LEICA SL2" text. So the clipping did not work in pdf_render. 

The technical drawings are also present thrice, but two copies are clipped (I think).

Another image (pdf_render, evice):
![image](https://user-images.githubusercontent.com/10325838/208741746-e8cf3008-450b-43b8-9878-a2ac84add780.png)

![image](https://user-images.githubusercontent.com/10325838/208741717-814a9159-85a0-48fc-ba76-14ffa7660f5e.png)

In this picture, I see two issues: Missing special characters (e.g. TM), and missing gray outlines of text hanging over the gray box. These outlines are created by "Stroke Text" (opposed to normal "Fill Text", at least that is term Master PDF uses).

![image](https://user-images.githubusercontent.com/10325838/208742930-9bfc3c00-5039-4401-a0bc-f2c9debe5a89.png)

The camera body is too dark (the texture is still there, but I have to increase brightness to see it). Also, there is a gray line around the left and top border of the image (not visible at every zoom level).

zoomed out (pdf_render, evice):

![image](https://user-images.githubusercontent.com/10325838/208743615-28d2c7bb-99eb-47b9-b1c9-5bd499a52210.png)

![image](https://user-images.githubusercontent.com/10325838/208743545-22ffcdec-b069-456b-8547-c75a129f5dd1.png)

I think when you zoom out far, you are supposed to render certain lines at a width of at least 1px per PDF specification.

**Comment by s3bk** [comment]
> I think when you zoom out far, you are supposed to render certain lines at a width of at least 1px per PDF specification.

Yes, but it makes no sense, so I did not implement it that way. It would make rendering a PDF depend on the resolution of the output...

> In this picture, I see two issues: Missing special characters (e.g. TM), and missing gray outlines of text hanging over the gray box. These outlines are created by "Stroke Text" (opposed to normal "Fill Text", at least that is term Master PDF uses).

If they use one of the standard fonts, then that is quite possible; and only they and Adobe are to blame for not distributing them freely.

And yes, the current clipping implementation is a hack at best. Unfortunately Pathfinder does not implement clipping, so it would need to be done with a geometry library.
